src/legacy/ui/public/vis/editors/default/components/default_editor_agg_params.tsx

```diff
@@ -0,0 +1,268 @@
+import {
+  getAggParamsToRender,
+  getError,
+  getAggTypeOptions,
+  ParamInstance,
+  isInvalidParamsTouched,
+} from './default_editor_agg_params_helper';
+import {
+  aggTypeReducer,
+  AGG_TYPE_ACTION_KEYS,
+  aggParamsReducer,
+  AGG_PARAMS_ACTION_KEYS,
+  initAggParamsState,
+  AggParamsItem,
+} from './default_editor_agg_params_state';
+import { editorConfigProviders } from '../../config/editor_config_providers';
+import { FixedParam, TimeIntervalParam, EditorParamConfig } from '../../config/types';
+// TODO: Below import is temporary, use `react-use` lib instead.
+// eslint-disable-next-line @kbn/eslint/no-restricted-paths
+import { useUnmount } from '../../../../../../../plugins/kibana_react/public/util/use_unmount';
+
+const FIXED_VALUE_PROP = 'fixedValue';
+const DEFAULT_PROP = 'default';
+type EditorParamConfigType = EditorParamConfig & {
+  [key: string]: unknown;
+};
+export interface SubAggParamsProp {
+  formIsTouched: boolean;
+  vis: Vis;
+  onAggParamsChange: (agg: AggParams, paramName: string, value: unknown) => void;
+  onAggTypeChange: (agg: AggConfig, aggType: AggType) => void;
+  onAggErrorChanged: (agg: AggConfig, error?: string) => void;
+}
+interface DefaultEditorAggParamsProps extends SubAggParamsProp {
+  agg: AggConfig;
+  aggIndex?: number;
+  aggIsTooLow?: boolean;
+  className?: string;
+  groupName: string;
+  indexPattern: IndexPattern;
+  responseValueAggs: AggConfig[] | null;
+  state: VisState;
+  setTouched: (isTouched: boolean) => void;
+  setValidity: (isValid: boolean) => void;
+}
+
+function DefaultEditorAggParams({
+  agg,
+  aggIndex = 0,
+  aggIsTooLow = false,
+  className,
+  groupName,
+  formIsTouched,
+  indexPattern,
+  responseValueAggs,
+  state = {} as VisState,
+  vis,
+  onAggParamsChange,
+  onAggTypeChange,
+  setTouched,
+  setValidity,
+  onAggErrorChanged,
+}: DefaultEditorAggParamsProps) {
+  const groupedAggTypeOptions = getAggTypeOptions(agg, indexPattern, groupName);
+  const errors = getError(agg, aggIsTooLow);
+
+  const editorConfig = editorConfigProviders.getConfigForAgg(
+    aggTypes.byType[groupName],
+    indexPattern,
+    agg
+  );
+  const params = getAggParamsToRender({ agg, editorConfig, responseValueAggs, state }, vis);
+  const allParams = [...params.basic, ...params.advanced];
+  const [paramsState, onChangeParamsState] = useReducer(
+    aggParamsReducer,
+    allParams,
+    initAggParamsState
+  );
+  const [aggType, onChangeAggType] = useReducer(aggTypeReducer, { touched: false, validity: true });
+
+  const isFormValid =
+    aggType.validity &&
+    Object.keys(paramsState).every((paramsName: string) => paramsState[paramsName].validity);
+  const isAllInvalidParamsTouched = isInvalidParamsTouched(agg.type, aggType, paramsState);
+
+  // reset validity before component destroyed
+  useUnmount(() => setValidity(true));
+
+  useEffect(
+    () => {
+      Object.keys(editorConfig).forEach(param => {
```

Comment:
```ts
Object.entries(editorConfig).forEach(([param, paramConfig]) => {
```

that way we iterate directly over all entries and don't need the line below that.

Reply:
Thanks, updated.